File: lab_11_langgraph/agents/writer-agent.js
import { ChatOpenAI } from "@langchain/openai";
import { HumanMessage, SystemMessage } from "@langchain/core/messages";

/**
 * Writer Agent – turns research notes into a clear, user-facing answer.
 * Used as a node in the LangGraph; reads shared state (userQuery, researchNotes), writes finalAnswer.
 */
export class WriterAgent {
  constructor(apiKey) {
    this.model = new ChatOpenAI({
      modelName: "openai/gpt-4o-mini",
      temperature: 0.5,
      configuration: { baseURL: "https://openrouter.ai/api/v1", apiKey: apiKey },
    });
    this.name = "WriterAgent";
  }

  /**
   * Node function: takes graph state, returns partial state update (finalAnswer).
   * @param {Object} state - Graph state with userQuery and researchNotes
   * @returns {Promise<{ finalAnswer: string }>}
   */
  async run(state) {
    const { userQuery, researchNotes } = state;
    if (!researchNotes) {
      return { finalAnswer: "No research available to write from." };
    }

    const systemPrompt = `You are a writer. Using the research notes provided, write a clear, friendly, and concise answer to the user's question. Do not add made-up facts; stick to the research. Use plain language.`;

    const userPrompt = `User question: ${userQuery || "N/A"}\n\nResearch notes:\n${researchNotes}\n\nWrite the final answer for the user.`;

    const messages = [
      new SystemMessage(systemPrompt),
      new HumanMessage(userPrompt),
    ];

    const response = await this.model.invoke(messages);
    const finalAnswer = typeof response.content === "string" ? response.content : String(response.content);

    return { finalAnswer };
  }

  getInfo() {
    return { name: this.name, role: "Turn research into a clear answer" };
  }
}

File: lab_11_langgraph_py/agents/writer_agent.py
"""
Writer Agent – turns research notes into a clear, user-facing answer.
Used as a node in the LangGraph; reads shared state (userQuery, researchNotes), writes finalAnswer.
"""
from langchain_openai import ChatOpenAI
from langchain_core.messages import HumanMessage, SystemMessage


class WriterAgent:
    def __init__(self, api_key: str) -> None:
        self.model = ChatOpenAI(
            model="openai/gpt-4o-mini",
            temperature=0.5,
            base_url="https://openrouter.ai/api/v1",
            api_key=api_key,
        )
        self.name = "WriterAgent"

    def run(self, state: dict) -> dict:
        """
        Node function: takes graph state, returns partial state update (final_answer).
        """
        user_query = state.get("userQuery") or state.get("user_query")
        research_notes = state.get("research_notes") or state.get("researchNotes")
        if not research_notes:
            return {"final_answer": "No research available to write from."}

        system_prompt = (
            "You are a writer. Using the research notes provided, write a clear, friendly, "
            "and concise answer to the user's question. Do not add made-up facts; stick to "
            "the research. Use plain language."
        )
        user_prompt = (
            f"User question: {user_query or 'N/A'}\n\n"
            f"Research notes:\n{research_notes}\n\n"
            "Write the final answer for the user."
        )
        messages = [
            SystemMessage(content=system_prompt),
            HumanMessage(content=user_prompt),
        ]

        response = self.model.invoke(messages)
        content = response.content
        final_answer = content if isinstance(content, str) else str(content)

        return {"final_answer": final_answer}

    def get_info(self) -> dict:
        return {"name": self.name, "role": "Turn research into a clear answer"}
